File: oop/05-interfaces/csharp/CashRegister.cs
/*
 * Demonstrates: Interfaces - Single Interface Implementation
 * Concept: CashRegister implements only ITransactable as it does not require audit logging
 */

namespace FinanceDomain.Interfaces;

public class CashRegister : ITransactable
{
    private decimal _balance;

    public string RegisterId { get; }
    public decimal Balance => _balance;

    public CashRegister(string registerId, decimal openingFloat)
    {
        RegisterId = registerId;
        _balance = openingFloat;
    }

    public void Deposit(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Amount must be greater than zero.");

        _balance += amount;
    }

    public void Withdraw(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Amount must be greater than zero.");

        if (amount > _balance)
            throw new InvalidOperationException($"Insufficient cash in register. Available: R{_balance:F2}");

        _balance -= amount;
    }

    public decimal CloseRegister()
    {
        decimal finalBalance = _balance;
        _balance = 0;
        return finalBalance;
    }
}

File: oop/05-interfaces/typescript/cashRegister.ts
/*
 * Demonstrates: Interfaces - Single Interface Implementation
 * Concept: CashRegister implements only ITransactable as it does not require audit logging
 */

import { ITransactable } from "./transactable";

export class CashRegister implements ITransactable {
    private _balance: number;

    public readonly registerId: string;

    constructor(registerId: string, openingFloat: number) {
        this.registerId = registerId;
        this._balance = openingFloat;
    }

    get balance(): number {
        return this._balance;
    }

    deposit(amount: number): void {
        if (amount <= 0) {
            throw new Error("Amount must be greater than zero.");
        }
        this._balance += amount;
    }

    withdraw(amount: number): void {
        if (amount <= 0) {
            throw new Error("Amount must be greater than zero.");
        }
        if (amount > this._balance) {
            throw new Error(`Insufficient cash in register. Available: R${this._balance.toFixed(2)}`);
        }
        this._balance -= amount;
    }

    closeRegister(): number {
        const finalBalance = this._balance;
        this._balance = 0;
        return finalBalance;
    }
}
